packages/tql/src/error.ts

```diff
@@ -0,0 +1,42 @@
+export class TqlError<T extends keyof typeof messages> extends Error {
+  constructor(
+    public code: T,
+    ...details: Parameters<(typeof messages)[T]>
+  ) {
+    // @ts-expect-error - This is super hard to type but it does work correctly
```

Comment:
I'm not a big fan of committing these to code unless absolutely necessary but I am acutely aware of how annoying this typing can be. An alternative that simplifies the typing but makes things slightly more verbose could be to create error class for each error type. For example:

```ts
export class TqlError extends Error {}
export class UntemplatedSqlCallTqlError extends TqlError {}
export class DialectMethodNotImplementedTqlError extends TqlError {
  constructor(method: string) {
    super(`The dialect you are using does not implement this method: ${method}`);
  } 
}
```

It also allows for super simple error discrimination for users:

```ts
import { DialiectMethodNotImplementedTqlError } from "@vercel/tql";

try {
  // ..
} catch(err) {
  if (err instanceof DialectMethodNotImplementedTqlError) {
    // .. do work
  }

  throw err;
}
```

Reply:
I actually think the ergonomics of subclasses are significantly _worse_ for the user in most cases, though I'm willing to be swayed. The main reason I've gone this route is because it makes this possible:

```ts
catch (e) {
  // as soon as you type the opening quote, TypeScript gives you
  // every possible error from this library, requiring no documentation
  if (e instanceof TqlError && e.code === 'untemplated_sql_call')
}

catch (e) {
  if (e instanceof TqlError) {
    switch (e.type) {
      case 'untemplated_sql_call':
        break;
      default:
        console.log('Impossible if exhaustive');
    }
  }
}
```

as opposed to:

```ts
catch (e) {
  if (e instanceof SomeErrorFromThisLibrary) {
    // I have to both know that this error type exists and import it,
    // for every kind of error I'd like to handle
    // which makes documentation hard and discoverability even harder
  }
}

catch (e) {
  // oops, there's actually no way to switch over error types at all, 
  //I'd have to import every class I want to check for and if/elseif them. 
  //There's no way to make sure I'm handling every error type, 
  // and TypeScript gives me no help in making sure things are exhaustive.
}
```

I'm not particularly worried about this comment, though I'll look into whether there's a way to fix it -- I just couldn't find one in the limited time I had!

Reply:
Ah we can still do best of both worlds! The super-class `TqlError` constructor can require a `code` parameter which the subclass `DialectMethodNotImplementedTqlError` can provide.

```ts
class TqlError extends Error {
	code: string;

	constructor(code: string, message: string) {
		super(message);
		this.code = code;
 	}
}

class DialectMethodNotImplementedTqlError extends TqlError {
  constructor() {
  	super("dialect_method_not_implemented", <message>)
  }
}
```

It's a matter of preference though I think. I generally prefer to discriminate errors using `instanceof`.

Reply:
Right, but the problem here is that TypeScript is still dumb about them:

```ts
if (tqlError.code === '') // typescript has no idea what these can be
```

and additionally, even if I do know the code of an error:

```ts
if (tqlError.code === 'illegal_query_recursion') {
  // typescript has no way to connect this back to "oh, then the error is actually a `TqlIllegalRecursionError`
}
```

The authorship experience is also somewhat better when using a single error class, as typing:

```ts
throw new TqlError('
```

is all you need to do to see the entire list of errors it's possible to throw from the library, as opposed to having to type `Error` and hope VSCode's intellisense offers to import the right one 😆